internal/auth/client_kerberos.go

```diff
@@ -0,0 +1,477 @@
+		if err != nil {
+			return nil, "", false, message.Wrap(
+				err,
+				message.EAuthKerberosVerificationFailed,
+				"Failed to marshal the AP_REP packet",
+			)
+		}
+
+		authCred, err := a.Authenticator.GetCredDelegation()
+		if err != nil {
+			k.client.logger.Info(message.Wrap(
+				err,
+				message.EAuthKerberosVerificationFailed,
+				"Failed to unmarshal the Authenticator delegation packet",
+			))
+			// Accept but no cred delegation
+			return mar2, k.username, false, nil
+		}
+
+		if authCred != nil && authCred.HasDelegation() {
+			var cred krbmsg.KRBCred
+			err = cred.Unmarshal(authCred.Deleg)
+			if err != nil {
+				k.client.logger.Info(message.Wrap(
+					err,
+					message.EAuthKerberosVerificationFailed,
+					"Failed to marshal the KRB_CRED packet",
+				))
+				return mar2, k.username, false, nil
+			}
+			err = cred.DecryptEncPart(ticket.DecryptedEncPart.Key)
+			if err != nil {
+				return nil, "", false, err
+			}
+
+			cacheCreds, err := cred.ToCredentials()
+			if err != nil {
+				return nil, "", false, err
+			}
+			cache := credentials.CCacheFromCredentials(cacheCreds)
+			mar, err := cache.Marshal()
+			if err != nil {
+				return nil, "", false, err
+			}
+			k.credentials = mar
+		}
+
+		return mar2, k.username, false, nil
+	}
+
+	return nil, "", false, fmt.Errorf("Invalid token")
+}
+
+const (
+	SSH_MSG_USERAUTH_REQUEST = 50
+)
+
+// GSSAPIMicField is described in RFC4462 Section 3.5
+type GSSAPIMicField struct {
+	// SessionIdentifier is a random string identifying the ssh connection
+	// (different from containerSSHs identifier)
+	SessionIdentifier string
+	// Request is the action that's being requested (50: SSH_MSG_USERAUTH_REQUEST)
+	Request byte
+	// UserName is the username that the user requests to log in as
+	UserName string
+	// Service is the service being used ('ssh-connection')
+	Service string
+	// Method is the authentication method in use ('gssapi-with-mic')
+	Method string
+}
+
+func (mic *GSSAPIMicField) unmarshal(b []byte) error {
+	err := ssh.Unmarshal(b, mic)
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *kerberosAuthContext) VerifyMIC(micField []byte, micToken []byte) error {
+	var t gssapi.MICToken
+	err := t.Unmarshal(micToken, false)
+	if err != nil {
+		return err
+	}
+	t.Payload = micField
+	verified, err := t.Verify(k.key, keyusage.GSSAPI_INITIATOR_SIGN)
+	if err != nil {
+		return err
+	}
+	if !verified {
+		return message.NewMessage(
+			message.EAuthKerberosVerificationFailed,
+			"Verify() returned unverified but no error",
+		)
+	}
+
+	// MIC is verified, but need to ensure usernames match
+	var field GSSAPIMicField
+	err = field.unmarshal(micField)
+	if err != nil {
+		return err
+	}
+
+	if field.Request != SSH_MSG_USERAUTH_REQUEST || field.Service != "ssh-connection" || field.Method != "gssapi-with-mic" {
+		return message.NewMessage(
+			message.EAuthKerberosVerificationFailed,
+			"Verify() returned unverified but no error",
```

Comment:
This error message does not seem to be very helpful. Maybe we should improve it to help with debugging?